src/Moq/ExpressionExtensions.cs

```diff
@@ -334,6 +334,9 @@ private static bool PartialMatcherAwareEval_ShouldEvaluate(Expression expression
 				case ExpressionType.Parameter:
 					return false;
 
+				case ExpressionType.Extension:
+					return !(expression is MatchExpression);
+
```

Comment:
Instead of this change, perhaps it would be better to add a check right inside `Evaluator` that recognizes irreducible `Extension` nodes.